spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java

```diff
@@ -0,0 +1,214 @@
+
+	private ApplicationContext applicationContext;
+
+	/**
+	 * Construct an instance with the provided lists of roles and lifecycles, which must be of equal length.
+	 * @param roles the roles.
+	 * @param lifecycles the lifecycles corresponding to the roles.
+	 */
+	public SmartLifecycleRoleController(List<String> roles, List<SmartLifecycle> lifecycles) {
+		Assert.notNull(roles, "'roles' cannot be null");
+		Assert.notNull(lifecycles, "'lifecycles' cannot be null");
+		Assert.isTrue(roles.size() == lifecycles.size(), "'roles' and 'lifecycles' must be the same lenght");
+		Iterator<SmartLifecycle> iterator = lifecycles.iterator();
+		for (String role : roles) {
+			SmartLifecycle lifecycle = iterator.next();
+			addLifecycleToRole(role, lifecycle);
+		}
+	}
+
+	/**
+	 * Construct an instance with the provided map of roles/instances.
+	 * @param lifcycles the {@link MultiValueMap} of beans in roles.
+	 */
+	public SmartLifecycleRoleController(MultiValueMap<String, SmartLifecycle> lifcycles) {
+		for (Entry<String, List<SmartLifecycle>> lifecyclesInRole : lifcycles.entrySet()) {
+			String role = lifecyclesInRole.getKey();
+			for (SmartLifecycle lifecycle : lifecyclesInRole.getValue()) {
+				addLifecycleToRole(role, lifecycle);
+			}
+		}
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;
+	}
+
+	/**
+	 * Add a {@link SmartLifecycle} to the role.
+	 * @param role the role.
+	 * @param lifecycle the {@link SmartLifecycle}.
+	 */
+	public void addLifecycleToRole(String role, SmartLifecycle lifecycle) {
+		this.lifecycles.add(role, lifecycle);
+	}
+
+	/**
+	 * Add a {@link SmartLifecycle} bean to the role using its name.
+	 * @param role the role.
+	 * @param lifecycleBeanName the bean name of the {@link SmartLifecycle}.
+	 */
+	public void addLifecycleToRole(String role, String lifecycleBeanName) {
+		Assert.state(this.applicationContext != null, "An application context is required to use this method");
+		this.lazyLifecycles.add(role, lifecycleBeanName);
+	}
+
+	/**
+	 * Add a {@link SmartLifecycle} beans to the role using their names.
+	 * @param role the role.
+	 * @param lifecycleBeanNames the bean names of the {@link SmartLifecycle}s.
+	 */
+	public void addLifecyclesToRole(String role, List<String> lifecycleBeanNames) {
+		Assert.state(this.applicationContext != null, "An application context is required to use this method");
+		for (String lifecycleBeanName : lifecycleBeanNames) {
+			this.lazyLifecycles.add(role, lifecycleBeanName);
+		}
+	}
+
+	/**
+	 * Start all registered {@link SmartLifecycle}s in the role.
+	 * @param role the role.
+	 */
+	public void startLifecyclesInRole(String role) {
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		List<SmartLifecycle> lifecycles = new ArrayList<SmartLifecycle>(this.lifecycles.get(role));
+		if (lifecycles != null) {
```

Comment:
`!lifecycles.isEmpty()`?
Because `new ArrayList` always produces a new instance.
From other side: does it really make sense to have a copy for the the `MultiMap.get()` result ?